keps/sig-architecture/2019-03-19-go-modules.md

```diff
@@ -101,12 +104,63 @@ In addition to simply keeping up with the go ecosystem, go modules provide many
+1. Generate `go.mod` files for `k8s.io/kubernetes` and each staging component (e.g. `k8s.io/client-go` and `k8s.io/api`) as a distinct go module
   
-    * Pin dependencies to the same versions that are currently listed in Godeps.json
+    * Add `require` and `replace` directives in all the `go.mod` files to register a preference for the same dependency versions currently listed in `Godeps.json`
 
-2. Change vendor management and verification scripts in `kubernetes/kubernetes` to use go module commands
-3. Change the staging component publishing bot to modify the `go.mod` pinned versions of peer components, rather than rewriting Godeps.json files
+        ```
+        require (
+          bitbucket.org/bertimus9/systemstat v0.0.0-20180207000608-0eeff89b0690
+          ...
+        )
+        replace bitbucket.org/bertimus9/systemstat => bitbucket.org/bertimus9/systemstat v0.0.0-20180207000608-0eeff89b0690
+        ...
+        ```
+
+    * Add `require` and `replace` directives in the `k8s.io/kubernetes` `go.mod` file to point to the staging component directories:
+
+        ```
+        require (
+          k8s.io/api v0.0.0
+          ...
+        )
+        replace k8s.io/api => ./staging/src/k8s.io/api
+        ...
+        ```
+
+    * Add `require` and `replace` directives in the staging component `go.mod` files to point to the peer component directories:
+
+        ```
+        require (
+          k8s.io/api v0.0.0
+          ...
+        )
+        replace k8s.io/api => ../api
+        ...
+        ```
+
+2. Change vendor creation and verification scripts in `kubernetes/kubernetes` to use go module commands to:
+
+    * Sync dependency versions between `k8s.io/kubernetes` and in staging component `go.mod` files
+    * Build the vendor directory (e.g. `rm -fr vendor; go mod vendor`)
+    * Generate the vendored `LICENSES` file
+
+3. Remove `Godeps.json` files from `kubernetes/kubernetes` and staging components.
+With the change to go modules, the only use for these is as a hint to non-module-based downstream consumers of the published staging components.
+
+4. Remove the custom `godep` fork from `kubernetes/kubernetes`
+
+See the [alternatives](#alternatives-to-vendoring-using-go-modules) section for other vendoring tools considered.
+
+### Publish staging component modules to individual repositories
+
+Update the staging component publishing bot:
+1. Rewrite the staging component `require` directives in `go.mod` files to require specific published versions (the same thing done today for `Godeps.json` files)
+2. Remove the staging component `replace` directives in `go.mod` files (relative path references don't make sense for independent repositories)
+3. Stop including `vendor` content in published modules. Vendor folders are ignored in non-top-level modules, and published `go.mod` files inform `go get` of desired versions of transitive dependencies.
```

Comment:
if/when our staging repos include commands, will we change our tune?

Reply:
that's a good question. I think it would depend on where canonical builds are being run from, and the requirements to have hermetic build environments, and the state/availability of mirror/cache proxies at that point in time to insulate against disappearing github repos, etc.